### PR TITLE
Mate proofs

### PR DESCRIPTION
**Pull request by robertnurnberg**
This is my first attempt to signal when a mate is proven on cdb, and help cdb come up with a proof.

Not sure about some details: have added those with `@vondele` in the code.

Also, two parts can be parallelized later I think. But wanted to get a working version right first.

### CODE REVIEW

**Comment by robertnurnberg** [comment]
By the way, once a mate is proven on cdb, the PV end is marked as "CHECKMATE". No other changes are visible to the user.

**Comment by robertnurnberg** [comment]
Sample output:
```
> python cdbsearch.py --epd "1r4k1/8/5PP1/K7/6NR/7B/1r6/7R w - -" --depthLimit 1
Searched epd :  1r4k1/8/5PP1/K7/6NR/7B/1r6/7R w - -
evalDecay    :  2
Concurrency  :  16
Starting date:  2023-05-28T22:25:28.877953
Search at depth  1
  cdb PV len:  5
  score     :  29995
  PV        :  g4h6 g8f8 g6g7 f8e8 g7g8q CHECKMATE
  PV len    :  5
  queryall  :  27
  bf        :  27.00
  inflight  :  1.19
  chessdbq  :  18
  enqueued  :  0
  unscored  :  0
  date      :  2023-05-28T22:25:31.244970
  total time:  0:00:02.36
  req. time :  131
  URL       :  https://chessdb.cn/queryc_en/?1r4k1/8/5PP1/K7/6NR/7B/1r6/7R_w_-_-_moves_g4h6_g8f8_g6g7_f8e8_g7g8q
  ```

**Comment by robertnurnberg** [comment]
So what I still plan to do in this PR:

1. Make `search` in `pv_has_proven_mate` use `evalDecay=0`. => Needs `search` to take the latter as a parameter.
2. Find smallest `depth` for `search` in `pv_has_proven_mate` to guarantee it finds terminal checkmate node.
3. Optimize what calls to run in parallel. (I think YES for `queryall` and NO for `search`, as the latter is recursive and likely returns `False` anyway, so wasted effort to also resolve all the siblings at the same time.)
4. Add "# X" for mate-in-X info if a PV with a proven mate has been found.
5. Add routine that checks if the full subtree of root has been scored, and  then terminate search. This will only be invoked if the PV ends in a terminal node, and will avoid useless continuation of exploration.
6. Deal with some left-over comments regarding executors.

I plan to update this comment as I go along.

File: cdbsearch.py [hunk 1]
```diff
@@ -294,7 +334,7 @@ async def search(self, board, depth):
         ):
             return 0, ["draw"]
 
-        # get current ranking, use an executor to limit total requests in flight
+        # get current ranking, use an executor to limit total requests in flight TODO: update this comment
```
Comment (robertnurnberg) [inline cdbsearch.py:337]: @vondele What would be a good comment here?

**Comment by robertnurnberg** [comment]
I think this works nicely. Some examples from the matetrack .epd:

```
> python cdbsearch.py --epd "2N2B2/2N1r3/8/3nQ2R/1k6/8/2B5/R3K3 w Q -" --depthLimit 1 --evalDecay 0
Root position:  2N2B2/2N1r3/8/3nQ2R/1k6/8/2B5/R3K3 w Q -
evalDecay    :  0
Concurrency  :  16
Starting date:  2023-05-29T17:02:42.981811
Search at depth  1
  cdb PV len:  3
  score     :  29997
  PV        :  e1c1 d5c7 e5c5 CHECKMATE (#2)
  PV len    :  3
  queryall  :  26
  bf        :  26.00
  inflight  :  1.81
  chessdbq  :  14
  enqueued  :  0
  unscored  :  0
  date      :  2023-05-29T17:02:44.177978
  total time:  0:00:01.19
  req. time :  85
  URL       :  https://chessdb.cn/queryc_en/?2N2B2/2N1r3/8/3nQ2R/1k6/8/2B5/R3K3_w_Q_-_moves_e1c1_d5c7_e5c5
```

```
> python cdbsearch.py --epd "3r4/3N2kr/1p6/pBpn1p2/Q2PR1p1/P7/1P4P1/2q3K1 w - -" --depthLimit 1 --evalDecay 0
Root position:  3r4/3N2kr/1p6/pBpn1p2/Q2PR1p1/P7/1P4P1/2q3K1 w - -
evalDecay    :  0
Concurrency  :  16
Starting date:  2023-05-29T17:01:02.851931
Search at depth  1
  cdb PV len:  10
  score     :  -29990
  PV        :  g1f2 f5e4 a4b3 h7h1 b3c3 d5c3 b5f1 c1f4 f2e1 f4e3 CHECKMATE (#-5)
  PV len    :  10
  queryall  :  12804
  bf        :  12804.00
  inflight  :  5.92
  chessdbq  :  5454
  enqueued  :  0
  unscored  :  0
  date      :  2023-05-29T17:04:01.924694
  total time:  0:02:59.07
  req. time :  32
  URL       :  https://chessdb.cn/queryc_en/?3r4/3N2kr/1p6/pBpn1p2/Q2PR1p1/P7/1P4P1/2q3K1_w_-_-_moves_g1f2_f5e4_a4b3_h7h1_b3c3_d5c3_b5f1_c1f4_f2e1_f4e3
```

And a check for a checkmate position:
```
> python cdbsearch.py --epd "8/5p1p/5P1k/5P1p/p4B1P/8/5p2/5K2 b - - 1 7"
Root position:  8/5p1p/5P1k/5P1p/p4B1P/8/5p2/5K2 b - - 1 7
evalDecay    :  2
Concurrency  :  16
Starting date:  2023-05-29T17:04:19.839895
Search at depth  1
  cdb PV len:  0
  score     :  -30000
  PV        :  CHECKMATE
  PV len    :  0
  URL       :  https://chessdb.cn/queryc_en/?8/5p1p/5P1k/5P1p/p4B1P/8/5p2/5K2_b_-_-_1_7
```

File: cdbsearch.py (hunk 1 above)
Comment (vondele) [reply cdbsearch.py:337]: just `# get current ranking` the other part is commented elsewhere (and currently not really right).

**Comment by vondele** [comment]
so one question, with proven mate, you use the common definition (i.e. a mate can be reached in no more moves than the found line)? 

This is not an attempt to find the shortest mate. Have you done a check on matetrack epd, and is it returning the expected answer in all cases ? 

I have one more question on a particular bit of code, I'll add it there.

File: cdbsearch.py [hunk 2]
```diff
@@ -498,7 +498,14 @@ async def cdbsearch(epd, depthLimit, concurrency, evalDecay, cursedWins=False):
             pvlen = len(pv) - 1
             if pv[-1] == "checkmate":
                 if await chessdb.pv_has_proven_mate(board.epd(), pv):
-                    pv[-1] = "CHECKMATE"
+                    pv[-1] = "CHECKMATE" + (
+                        ""
+                        if pvlen == 0
+                        else f" (#{(pvlen+1)//2})"
```
Comment (vondele) [inline cdbsearch.py:504]: would it make sense to show this at the score line instead?

File: cdbsearch.py [hunk 3]
```diff
@@ -137,6 +131,52 @@ async def add_cdb_pv_positions(self, epd):
...
+
+        # we need to check if the _given_ PV is a correct mating line:
+        for m in pv[:2]:
+            board.push(chess.Move.from_uci(m))
+        if not await self.pv_has_proven_mate(board.epd(), pv[2:]):
+            return False
+        for _ in [0, 1]:
+            board.pop()
+
+        for m in scored_db_moves:
```
Comment (vondele) [inline cdbsearch.py:166]: This for loop should be commented, it is not quite clear to me what it does.

**Comment by robertnurnberg** [comment]
Not sure I get your question. With proven mate I mean that if the script shows CHECKMATE (#4) then even with best play the mated side cannot avoid being mated after at most 4 moves.

And yes, it is not an attempt to find the shortest mate. Although the hope would be that exploration with cdbsearch will eventually find the shortest mate as well. Just cannot prove that it is the shortest one.

File: cdbsearch.py (hunk 3 above)
Comment (robertnurnberg) [reply cdbsearch.py:166]: I had exactly the same thought. But then I thought it sits maybe nicer next to CHECKMATE, and if people use scripts or some such to data mine cdbsearch output they can always be sure it `eval` a numerical value.

File: cdbsearch.py (hunk 2 above)
Comment (robertnurnberg) [reply cdbsearch.py:504]: I had exactly the same thought. But then I thought it sits maybe nicer next to CHECKMATE, and if people use scripts or some such to data mine cdbsearch output they can always be sure it `eval` a numerical value.

File: cdbsearch.py (hunk 3 above)
Comment (robertnurnberg) [reply cdbsearch.py:166]: python allows _ to be used for variable whose value is irrelevant. frequently used in for loops where you just need to run a number of times

**Comment by robertnurnberg** [comment]
So I think this just leaves 6. in https://github.com/vondele/cdbexplore/pull/42#issuecomment-1566676461, which we could also do in a future PR.

And see https://github.com/vondele/cdbexplore/pull/42#issuecomment-1567253432 for some tests with positions from the matetrack problem suite.

This may be ready for merge, and then we can ask e.g. David on discord to give the new script a go.

File: cdbsearch.py (hunk 3 above)
Comment (robertnurnberg) [reply cdbsearch.py:167]: So this loop will only be reached if all the moves have a score. So these are alternative defences for the to-be-mated side. We need to check that they are all valid mate PV's. We do this recursively, always checking just the defenders moves.

Comment (robertnurnberg) [reply cdbsearch.py:167]: suggestion for a comment:
```
# now we check if all the currently non-best moves also inevitably lead to the defender being mated
```

**Comment by robertnurnberg** [comment]
Sorry, I am putting this on hold again. For the call `python cdbsearch.py --epd "2r5/1p6/1Nk1p1pn/QpPp1p1r/6p1/nN3b1q/3R4/2BR2K1 w - -"` the user needs to wait days (!) for some output after this:
```
> python cdbsearch.py --epd "2r5/1p6/1Nk1p1pn/QpPp1p1r/6p1/nN3b1q/3R4/2BR2K1 w - -"
Root position:  2r5/1p6/1Nk1p1pn/QpPp1p1r/6p1/nN3b1q/3R4/2BR2K1 w - -
evalDecay    :  2
Concurrency  :  16
Starting date:  2023-05-29T22:01:09.424925
Search at depth  1
  cdb PV len:  49
```

For comparison, main/master gives this:
```
> python cdbsearch.py --epd "2r5/1p6/1Nk1p1pn/QpPp1p1r/6p1/nN3b1q/3R4/2BR2K1 w - -"
Searched epd :  2r5/1p6/1Nk1p1pn/QpPp1p1r/6p1/nN3b1q/3R4/2BR2K1 w - -
evalDecay    :  2
Concurrency  :  16
Starting date:  2023-05-29T22:02:34.872553
Search at depth  1
  cdb PV len:  49
  score     :  29951
  PV        :  b3d4 c6c7 d4b5 a3b5 b6d5 c7b8 a5a8 b8a8 d2a2 a8b8 c1f4 e6e5 f4e5 b5d6 e5d6 c8c7 d6c7 b8c8 a2a8 c8d7 a8d8 d7c6 d8d6 c6b5 d1b1 b5a4 b1a1 a4b3 d6b6 b3c2 a1a2 c2d3 d5f4 d3e3 a2a3 e3d4 a3a4 d4c3 c7e5 c3d2 b6b2 d2e3 b2b3 e3d2 a4a2 d2e1 e5c3 e1d1 b3b1 checkmate
  PV len    :  49
  queryall  :  105
  bf        :  105.00
  inflight  :  8.14
  chessdbq  :  58
  enqueued  :  0
  unscored  :  0
  date      :  2023-05-29T22:02:37.494058
  total time:  0:00:02.62
  req. time :  45
  URL       :  https://chessdb.cn/queryc_en/?2r5/1p6/1Nk1p1pn/QpPp1p1r/6p1/nN3b1q/3R4/2BR2K1_w_-_-_moves_b3d4_c6c7_d4b5_a3b5_b6d5_c7b8_a5a8_b8a8_d2a2_a8b8_c1f4_e6e5_f4e5_b5d6_e5d6_c8c7_d6c7_b8c8_a2a8_c8d7_a8d8_d7c6_d8d6_c6b5_d1b1_b5a4_b1a1_a4b3_d6b6_b3c2_a1a2_c2d3_d5f4_d3e3_a2a3_e3d4_a3a4_d4c3_c7e5_c3d2_b6b2_d2e3_b2b3_e3d2_a4a2_d2e1_e5c3_e1d1_b3b1

Search at depth  2
  cdb PV len:  49
  score     :  29951
  PV        :  b3d4 c6c7 d4b5 a3b5 b6d5 c7b8 a5a8 b8a8 d2a2 a8b8 c1f4 e6e5 f4e5 b5d6 e5d6 c8c7 d6c7 b8c8 a2a8 c8d7 a8d8 d7c6 d8d6 c6b5 d1b1 b5a4 b1a1 a4b3 d6b6 b3c2 a1a2 c2d3 d5f4 d3e3 a2a3 e3d4 a3a4 d4c3 c7e5 c3d2 b6b2 d2e3 b2b3 e3d2 a4a2 d2e1 e5c3 e1d1 b3b1 checkmate
  PV len    :  49
  queryall  :  217
  bf        :  14.73
  inflight  :  3.94
  chessdbq  :  65
  enqueued  :  0
  unscored  :  0
  date      :  2023-05-29T22:02:38.242267
  total time:  0:00:03.36
  req. time :  51
  URL       :  https://chessdb.cn/queryc_en/?2r5/1p6/1Nk1p1pn/QpPp1p1r/6p1/nN3b1q/3R4/2BR2K1_w_-_-_moves_b3d4_c6c7_d4b5_a3b5_b6d5_c7b8_a5a8_b8a8_d2a2_a8b8_c1f4_e6e5_f4e5_b5d6_e5d6_c8c7_d6c7_b8c8_a2a8_c8d7_a8d8_d7c6_d8d6_c6b5_d1b1_b5a4_b1a1_a4b3_d6b6_b3c2_a1a2_c2d3_d5f4_d3e3_a2a3_e3d4_a3a4_d4c3_c7e5_c3d2_b6b2_d2e3_b2b3_e3d2_a4a2_d2e1_e5c3_e1d1_b3b1
```

So I think some output for the user is need before/during attempted mate proving. Will look into that tomorrow.

**Comment by robertnurnberg** [comment]
So my current working assumption is that the massive initial delays seen in https://github.com/vondele/cdbexplore/pull/42#issuecomment-1567476257 are due to some sort of `search` (!) explosions within the mate proving, in part likely due to the generous extensions along cdb PV lines.

But in fact, the `search` calls within `pv_has_proven_mate` only serve a single purpose: to extract the PV for the epd up to length `len(pv) - 2`. So there are two solutions:

1. either create a purpose built function like `extract_PV`
2. or make `search` take a parameter like `extractPVonly` that avoids all the bells and whistles of normal search

the disadvantage of 1. is a lot of code repitition with `search`, while what is not nice about 2. is that `search` gets more convoluted. However, I currently prefer 2. Agreed? @vondele 

On top of that, I would like to change the print output behaviour before mate proving starts, so the user gets warned. We should defer merging until these two issues are addressed I think.

**Comment by vondele** [comment]
BTW, that search was actually what I asked to comment. I wonder how the search returning the PV can explode, in principle it is just linear (i.e. can't explode).... except for those cases where moves are degenerate in evaluation. For extracting 'the' PV, it would be enough to just pick one of the degenerate moves. I guess that could be an option for the move_depth function ?

I wonder if the PV line could just be queried from cdb instead of retrieved via a search? If the main purpose of the function is to show that cdb has proven the mate, getting the PV from there might work?

**Comment by vondele** [comment]
concerning the output, I think that adding a line of output is probably no issue, but I believe one should put the mate proving behind a command line flag, and make it optional. I suspect that proving mate for any mate that is a bit deeper could take very long, because it will typically have a very large branching factor.

**Comment by robertnurnberg** [comment]
> BTW, that search was actually what I asked to comment. I wonder how the search returning the PV can explode, in principle it is just linear (i.e. can't explode).... except for those cases where moves are degenerate in evaluation. For extracting 'the' PV, it would be enough to just pick one of the degenerate moves. I guess that could be an option for the move_depth function ?

Maybe explosion is the wrong word. Maybe it's just purely the size of the problem that generates these long waits. But I would like to at least try to reduce those waiting times. And I believe replacing these `search` calls by a simpler mechanism would help.

Not sure I get what you mean by 'degenerate' moves and doing it via the `move_depth` function.

With 'the' PV I mean for each node to pick the/a best move for as long as possible, or for as long as requested. As part of (successful) mate proving these should always end in a `checkmate` node.

> I wonder if the PV line could just be queried from cdb instead of retrieved via a search? If the main purpose of the function is to show that cdb has proven the mate, getting the PV from there might work?

I do not think using `querypv` would be OK here, but I am not sure. We must find the PV within the local search tree, as that is where we construct the proof. In theory there could be sync-delays with cdb? Or cdb could have found a better mate in the meantime, with new nodes not yet in local search tree? In any case, finding the PV line in the cached tree locally may be faster than a single API call, even for long PVs?

**Comment by robertnurnberg** [comment]
> concerning the output, I think that adding a line of output is probably no issue, but I believe one should put the mate proving behind a command line flag, and make it optional. I suspect that proving mate for any mate that is a bit deeper could take very long, because it will typically have a very large branching factor.

I thought of something more subtle: like printing the whole PV apart from the last word. Then go into checking and print the final word once checked. Would mean no change for PVs without `checkmate`, and signal to the user when the latter are being proved. OK?

Happy to include a command line flag. But in what scenario would users want to run without it? Even if without the flag a shorter mate is found faster (as less time was "wasted" on trying to prove a sub-optimal mate), the user will never know whether they can trust the report mate score.

> A long wait time is no problem if the user is doing a deliberate mate search, but e.g. if analyzing a played game which ends in mate, this could become a problem.

Hm, you mean as part of `cdbbulksearch` for a full .pgn? I get what you mean now. Then how about this: we add a flat `--proveMates` to both scripts, but make it default to `True` in `cdbsearch` and default to `False` in the bulk version. OK?

**Comment by vondele** [comment]
with degenerate moves I mean two (best) moves with the same score. When looking for the PV, it would be sufficient to follow one of the two lines, while currently search will search both moves.

I think that just searching the local cache of positions for a PV line would be very fast indeed, much faster than an api call.

**Comment by vondele** [comment]
`--proveMates` option would work indeed. I would prefer to have it false by default for both scripts for consistency, if we find it works well and in reasonable time we could still switch it to default to true. Right now, I'm somewhat concerned it could take a very long time for all but the shortest mates.

**Comment by robertnurnberg** [comment]
Agreed about 'degenerate' moves, and the speed re API. 

Picking just one of two degenerate moves would mean even more tweaking of `search` in case it gets passed the flag `extractPVonly`. So I think writing a new stand-alone routine is best at this stage, maybe called `extract_cached_PV`. I could try to sub-contract some of the overlapping work from `search`, to avoid code duplication. Agreed?

**Comment by vondele** [comment]
yes, having an 'extract_cached_PV' would be useful. If it works well there might be other uses for it.

**Comment by robertnurnberg** [comment]
Text from my commit message in branch on which this PR is based:

> This implements all the discussed changes:
> * introduce `--proveMates` command line parameter
> * indicate in output when mate proving is going on
> * replace `search` in mate proving with dedicated `extract_cached_PV` method
> * use same parallelism in mate proving
> 
> The only thing I am a bit unsure about is whether to base `extract_cached_PV` on a parameter `board` or `epd`. In the end I went for the latter, as it may be faster and is enough for our needs for now.


I think some testing would be good. Early signs are encouraging. E.g. what took 3 minutes above now takes 1:40mins (and far fewer queryall request: 12804 vs 2260):
```
> python cdbsearch.py --epd "3r4/3N2kr/1p6/pBpn1p2/Q2PR1p1/P7/1P4P1/2q3K1 w - -" --depthLimit 1 --evalDecay 0 --proveMates
Root position:  3r4/3N2kr/1p6/pBpn1p2/Q2PR1p1/P7/1P4P1/2q3K1 w - -
evalDecay    :  0
Concurrency  :  16
Prove Mates  :  True
Starting date:  2023-05-30T19:32:34.693678
Search at depth  1
  cdb PV len:  10
  score     :  -29990
  PV        :  b5f1 g4g3 a4c4 h7h1 g1h1 c1h6 e4h4 h6h4 h1g1 h4h2 CHECKMATE (#-5)
  PV len    :  10
  queryall  :  2260
  bf        :  2260.00
  inflight  :  8.35
  chessdbq  :  1655
  enqueued  :  0
  unscored  :  0
  date      :  2023-05-30T19:34:13.796467
  total time:  0:01:39.10
  req. time :  59
  URL       :  https://chessdb.cn/queryc_en/?3r4/3N2kr/1p6/pBpn1p2/Q2PR1p1/P7/1P4P1/2q3K1_w_-_-_moves_b5f1_g4g3_a4c4_h7h1_g1h1_c1h6_e4h4_h6h4_h1g1_h4h2
```

**Comment by vondele** [comment]
one nit comment, `extract_cached_PV` is probably not the best name, as it is based on query_all, so it will go for a call to cdb if it is not in the cache. 

Somewhat more important is probably the fact that I think prove_mate_pv should be based on a board, not on an epd. For example, if extract_cached_PV returns a PV that contains a 3-fold but still ends in a mate position after the expected number of moves, it would in the current implementation probably still consider that a correct mating line? Since search won't return such PVs, it would probably be possible to implement this in extract_cached_PV, but it seems somewhat safer to do it in prove_mate_pv.

**Comment by robertnurnberg** [comment]
> one nit comment, `extract_cached_PV` is probably not the best name, as it is based on query_all, so it will go for a call to cdb if it is not in the cache.

We could go directly to the TT, without going through `queryall`? I think the only reason I went via `queryall` is to make it also work (in future) if we ask for PV for nodes that are not yet locally explored. Then name could stay. 

But you would prefer to keep implementation using `queryall` and change name? To `obtain_PV` ?

> Somewhat more important is probably the fact that I think prove_mate_pv should be based on a board, not on an epd. For example, if extract_cached_PV returns a PV that contains a 3-fold but still ends in a mate position after the expected number of moves, it would in the current implementation probably still consider that a correct mating line? Since search won't return such PVs, it would probably be possible to implement this in extract_cached_PV, but it seems somewhat safer to do it in prove_mate_pv.

That is a good point. I only thought of the main line, where this cannot open, as it comes from `search`. But when we check the alternative defences, this may in fact happen. Well spotted. Will try to fix it.

**Comment by robertnurnberg** [comment]
I have done the requested changes. New sample output:

```
> python cdbsearch.py --epd "3r4/3N2kr/1p6/pBpn1p2/Q2PR1p1/P7/1P4P1/2q3K1 w - -" --depthLimit 1 --evalDecay 0 --proveMates
Root position:  3r4/3N2kr/1p6/pBpn1p2/Q2PR1p1/P7/1P4P1/2q3K1 w - -
evalDecay    :  0
Concurrency  :  16
Prove Mates  :  True
Starting date:  2023-05-30T21:49:47.014405
Search at depth  1
  cdb PV len:  10
  score     :  -29990
  PV        :  g1f2 f5e4 a4b3 h7h1 b3c3 d5c3 b5e2 c1f4 e2f3 g4g3 CHECKMATE (#-5)
  PV len    :  10
  queryall  :  1910
  bf        :  1910.00
  inflight  :  7.15
  chessdbq  :  1465
  enqueued  :  0
  unscored  :  0
  date      :  2023-05-30T21:51:47.455372
  total time:  0:02:00.44
  req. time :  82
  URL       :  https://chessdb.cn/queryc_en/?3r4/3N2kr/1p6/pBpn1p2/Q2PR1p1/P7/1P4P1/2q3K1_w_-_-_moves_g1f2_f5e4_a4b3_h7h1_b3c3_d5c3_b5e2_c1f4_e2f3_g4g3
```

I have also added more comments within the routine `pv_has_proven_mate`, hoping the algorithm is now easier to follow.

**Comment by vondele** [comment]
`python cdbbulksearch.py --depthLimit 1 --evalDecay 0 --proveMates --concurrency 128 --bulkConcurrency 1 ../matetrack/ChestUCI_23102018.epd` leads to an error.

**Comment by vondele** [comment]
I'm getting a wrong answer for `python cdbsearch.py --epd "5K2/8/2qk4/2nPp3/3r4/6B1/B7/3R4 w - e6" --depthLimit 1 --evalDecay 0 --proveMates`
probably 
```diff
diff --git a/cdbsearch.py b/cdbsearch.py
index 778b2c7..d39d758 100644
--- a/cdbsearch.py
+++ b/cdbsearch.py
@@ -533,7 +533,7 @@ async def cdbsearch(
         print("  score     : ", bestscore)
         pvlen = len(pv) - 1 if pv[-1] in ["checkmate", "draw", "invalid"] else len(pv)
         print("  PV        : ", " ".join(pv[:-1]), end=" ", flush=True)
-        if args.proveMates and pv[-1] == "checkmate" and pvlen > 1:
+        if proveMates and pv[-1] == "checkmate" and pvlen >= 1:
             if await chessdb.pv_has_proven_mate(board.copy(), pv):
                 pv[-1] = "CHECKMATE" + (
                     f" (#{(pvlen+1)//2})" if bestscore > 0 else f" (#-{pvlen//2})"
```

**Comment by robertnurnberg** [comment]
Thanks for spotting these. The `args.proveMates` bug still worked, because this variable was defined _somewhere_ in the source file. As regards the `pvlen` error: I should have tested for #1! ;)

**Comment by robertnurnberg** [comment]
Ready to be merged, if your tests did not show up any other problems. :)

**Comment by vondele** [comment]
No further problems seen with your patch, I'll merge. 

(I did find a problem with memory use of the current version, I think I need to put back in place some safeguards I removed when moving to asyncio).

**Comment by robertnurnberg** [comment]
Just for later reference: I am seeing a massive speed-up with the most recent version, I guess because the checks in the mate proving are now done in parallel for all defending moves:

```
Root position:  2r5/1p6/1Nk1p1pn/QpPp1p1r/6p1/nN3b1q/3R4/2BR2K1 w - -
evalDecay    :  2
Concurrency  :  16
Prove Mates  :  True
Starting date:  2023-05-31T08:02:26.975750
Search at depth  1
  cdb PV len:  49
  score     :  29951
  PV        :  b3d4 c6c7 d4b5 a3b5 b6d5 c7b8 a5a8 b8a8 d2a2 a8b8 c1f4 e6e5 f4e5 b5d6 e5d6 c8c7 d6c7 b8c8 a2a8 c8d7 a8d8 d7c6 d8d6 c6b5 d1b1 b5a4 b1a1 a4b3 d6b6 b3c2 a1a2 c2d3 d5f4 d3e3 a2a3 e3d4 a3a4 d4c3 c7e5 c3d2 b6b2 d2e3 b2b3 e3d2 a4a2 d2c1 f4d3 c1d1 b3b1 checkmate
  PV len    :  49
  queryall  :  1305
  bf        :  1305.00
  inflight  :  1.07
  chessdbq  :  630
  enqueued  :  30
  unscored  :  30
  date      :  2023-05-31T08:04:49.925345
  total time:  0:02:22.94
  req. time :  226
  URL       :  https://chessdb.cn/queryc_en/?2r5/1p6/1Nk1p1pn/QpPp1p1r/6p1/nN3b1q/3R4/2BR2K1_w_-_-_moves_b3d4_c6c7_d4b5_a3b5_b6d5_c7b8_a5a8_b8a8_d2a2_a8b8_c1f4_e6e5_f4e5_b5d6_e5d6_c8c7_d6c7_b8c8_a2a8_c8d7_a8d8_d7c6_d8d6_c6b5_d1b1_b5a4_b1a1_a4b3_d6b6_b3c2_a1a2_c2d3_d5f4_d3e3_a2a3_e3d4_a3a4_d4c3_c7e5_c3d2_b6b2_d2e3_b2b3_e3d2_a4a2_d2c1_f4d3_c1d1_b3b1
  ```

Compare this with https://github.com/vondele/cdbexplore/pull/42#issuecomment-1567476257: 2 minutes  v 12 hours!

**Comment by vondele** [comment]
or cdb learned a few positions ... in that earlier message you had
```unscored  :  64425```
now
```unscored  :  30```
that of course makes a massive difference in run time.

**Comment by robertnurnberg** [comment]
Indeed, that's a very good point.

I'm quite excited now to see if and when we can prove *that* mate in 25 on cdb. :)

**Pull request merged**